### ISSUE TITLE
`enum.nonmember` type-decays Flag values

### ISSUE THREAD

**Issue by thejcannon** [python/cpython#120361]
# Bug report

### Bug description:

```pycon
>>> from enum import Flag, auto, nonmember
>>> 
>>> class DepType(Flag):
...     LIBRARY = auto()
...     SCRIPT = auto()
...     TEST = auto()
...     ALL = nonmember(LIBRARY | SCRIPT | TEST)
...  
>>> DepType.ALL
7
>>> type(DepType.ALL)
<class 'int'>
>>> DepType.LIBRARY | DepType.SCRIPT | DepType.TEST
<DepType.LIBRARY|SCRIPT|TEST: 7>
>>> type(DepType.LIBRARY | DepType.SCRIPT | DepType.TEST)
<flag 'DepType'>
```

this manifests by seeing a funky error on:
```pycon
>>> DepType.LIBRARY in DepType.ALL
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: argument of type 'int' is not iterable
```

If this isn't a bug, it should at least be mentioned on the docs.

### CPython versions tested on:

3.11

### Operating systems tested on:

Linux

### Linked PRs
* gh-120364
* gh-120511
* gh-120512

**Comment by thejcannon** [comment]
I don't see this functionality tested one way or another in `cpython/Lib/test/test_enum.py` so odds of it being undesired behavior have increased

**Comment by sobolevn** [comment]
This does not look like a bug to me. It does what is advertised: `ALL` is not an enum member. But, why is it an `int`?

Because `TEST` / `LIBRARY` / etc are converted to enum values **after** class body is evaluated. Inside the enum's body they are still just ints. Proof:

```python
>>> class DepType(Flag):
...     LIBRARY = auto()
...     TEST = auto()
...     print(LIBRARY, TEST, LIBRARY | TEST)
...     
1 2 3
```

So, `ALL` will just record their value and store it as a property.

I recommend doing something like: 

```python
>>> DepType.LIBRARY in DepType
True
```

This won't require to define `ALL` for this case.

But, I agree that adding a test case for this corner case might be a good idea.

**Comment by thejcannon** [comment]
> Because TEST / LIBRARY / etc are converted to enum values after class body is evaluated

Oh duh!

> This won't require to define ALL for this case.

The real-world scenario is a bit more hoop-y:

```python
def foo(dep_type: DepType = DepType.ALL):
    if DepType.LIBRARY in dep_type:
        do_thing()
```

So having the nice `.ALL` shorthand is great.

**Comment by ethanfurman** [comment]
Why are you having `.ALL` not be a member?  (Technically, it would be an alias.)

**Comment by sobolevn** [comment]
But, the code is not correct :)

```python
def foo(dep_type: DepType = DepType.ALL):
    if DepType.LIBRARY in dep_type:
        do_thing()
```

We now know that `DepType.ALL` is `int`. But, it is annotated as `DepType`. It should be:

```python
def foo(dep_type: DepType | int = DepType.ALL):
    if not isinstance(dep_typ, DepType):
        dep_type = DepType(dep_typ) 
    if DepType.LIBRARY in dep_type:
        do_thing()
```

And it will do what you want :)

**Comment by thejcannon** [comment]
Well in this case I want to be a nice easy `DepType` (I understand there's nuance here, but the point was that I don't want it to be a `int` (hence the bug report).

I think this might capture the intent best (albeit a little verbose):

```python
class DepType(Flag):
    LIBRARY = auto()
    SCRIPT = auto()
    TEST = auto()

    ALL: "DepType"

DepType.ALL = DepType.LIBRARY | DepType.SCRIPT | DepType.TEST
```

**Comment by ethanfurman** [comment]
@thejcannon You're last example is exactly the same as if you had not used `nonmember` in the first place.

**Comment by thejcannon** [comment]
> @thejcannon You're last example is exactly the same as if you had not used `nonmember` in the first place.

Not quite (see below) although:

> Why are you having `.ALL` not be a member? (Technically, it would be an alias.)

Welp, I incorrectly assumed it'd show up in `list(DepType)` as it is a member. But I just tried it and I'm wrong.

Apparently my intuiting surrounding `Flag` enums is just ALL out of wack 😥 .

```python
from enum import Flag, auto, nonmember


class DepType(Flag):
    LIBRARY = auto()
    SCRIPT = auto()
    TEST = auto()

    IMMEDIATE = nonmember(LIBRARY | SCRIPT | TEST)
    DELAYED: "DepType"
    ALIAS = LIBRARY | SCRIPT | TEST


DepType.DELAYED = DepType.LIBRARY | DepType.SCRIPT | DepType.TEST

print(type(DepType.IMMEDIATE))
print(type(DepType.DELAYED))
print(type(DepType.ALIAS))
print(list(DepType))
```

So, looks like I do just want `ALL = LIBRARY | SCRIPT | TEST` and everything "Just Works ™️"

I'll leave bug report open for the added test cases, and if anyone wants to noodle if the docs could be more illuminating.

Otherwise, pack it up nothing to see here. Just bad intuition.

**Comment by ethanfurman** [comment]
@thejcannon When `Flag` first came out, aliases were included in iterations -- that changed in 3.10 or 3.11.

**Comment by sobolevn** [comment]
Thanks everyone!
